### ISSUE TITLE
Log globbing anomalies

### ISSUE THREAD

**Issue by cdmihai** [Microsoft/msbuild#3642]
- Drive enumeration. The more general problem is when the glob pattern contains properties which evaluate to empty. (`$(src)/**/*.cs` evaluates to `/**/*.cs`).
- Original wildcard pattern is retuned ([as it sometimes happens](https://github.com/Microsoft/msbuild/blob/master/documentation/WhenGlobbingReturnsOriginalFilespec.md)). Maybe also include the reason.

**Comment by rainersigwald** [comment]
Let's spike on this in the next sprint: if it's easy to do (because we have a logging context where we'd like to collect the timing/bad patterns and log them) we'll do it, and if not then we'll describe what does exist here.

**Comment by Forgind** [comment]
I have two partial solutions (one per commit) [here](https://github.com/Forgind/msbuild/tree/exp-globbing-empty-property). Do you think it's better to try to sense a globbing anomaly while expanding a property or wait until we're globbing and assume no one wants to enumerate their whole drive? I could also try to pass information about empty properties from one to the other, but that would be complicated and a lot of extra information, so I don't think it's worth it.

**Comment by cdmihai** [comment]
Keep in mind these things (property evaluation, glob parsing/expanding) are on the hot path, so newly added logic (like additional regex matching) has a good chance of being picked up by RPS. For globbing anomalies I would just change the globbing code to record diagnostics and bubble them upwards to the evaluator / expander which can log appropriate message. For drive enumeration I would avoid duplicating the knowledge of glob syntax between the glob parsing code and the expander (glob parsing code should be the only one that knows the syntax). The glob parsing code should just record a "drive enumeration" diagnostic when it detects that the fixed directory part is a drive.

**Comment by rokonec** [comment]
@ladipro has any of your recent changes in globs addressed that?

**Comment by ladipro** [comment]
@rokonec no, the recent globbing changes did not improve logging. The unintended drive enumeration problem was somewhat alleviated in #5669 but I agree that this issue should be in backlog.

**Comment by mruxmohan-msft** [comment]
Are there any updates on logging/surfacing an exception for max path during wildcard expansion?

**Comment by ladipro** [comment]
@mruxmohan-msft there have been some internal discussions but no conclusion yet. The issue still exists. I'll leave a comment in #7029.

**Comment by rokonec** [comment]
@JanKrivanek This might be good candidate for analyzers V1, especially max path, it is so serious that it maybe shall be warning but warnings could break some builds
